### PR TITLE
Refactor GenerateDonutCatalog*.py tasks 

### PR DESCRIPTION
**Pull request by jbkalmbach**

### CODE REVIEW

File: python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py [hunk 1]
```diff
@@ -19,132 +19,194 @@
...
+        and pointing information.
+
+        Parameters
+        ----------
+        dataIds : list
+            List of the deferred dataset references pointing to the pieces
+            of the reference catalog we want in the butler.
+        refCats : list
+            List of the dataIds for the reference catalog shards.
+        """
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py]: Please help to add the doc string of `**kwargs` in `Parameters` as well. Thanks!

File: python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py [hunk 2]
```diff
@@ -19,132 +19,194 @@
...
 
-        # The filter in the reference catalog we want to use to find sources.
-        self.filterName = self.config.filterName
+        pipeBase.Task.__init__(self, **kwargs)
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py]: Are you trying to do the `super().__init__()` actually? If so, you could do the following instead:
```
super().__init__(**kwargs)
```

File: python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py [hunk 3]
```diff
@@ -19,132 +19,194 @@
...
+        refConfig = self.config.refObjLoader
+        self.refObjLoader = ReferenceObjectLoader(
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py:88]: Please give the comment to explain this attribute: `self.refObjLoader`. Thanks!

File: python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py [hunk 4]
```diff
@@ -19,132 +19,194 @@
...
-    def __init__(self, **kwargs):
-        super().__init__(**kwargs)
+    def __init__(self, dataIds, refCats, **kwargs):
+        """
+        Construct a source catalog from reference catalogs
+        and pointing information.
+
+        Parameters
+        ----------
+        dataIds : list
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py]: dataIds : list --> dataIds : `list`

File: python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py [hunk 5]
```diff
@@ -19,132 +19,194 @@
...
+            List of the deferred dataset references pointing to the pieces
+            of the reference catalog we want in the butler.
+        refCats : list
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py]: refCats : list --> refCats : `list`

File: python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py [hunk 6]
```diff
@@ -19,132 +19,194 @@
...
-        donutCatalogList = []
+        if self.config.doReferenceSelection:
+            goodSources = self.referenceSelector.selectSources(refCat)
+            refSelected = goodSources.selected
+        else:
+            refSelected = np.ones(len(refCat), dtype=bool)
 
-        for detector in camera:
+        if self.config.doDonutSelection:
+            print("Running Donut Selector")
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py]: Do you have something like the `self.log` in this child task class to use? If not, never mind. I just thought the DM team like to use the `logging` library everywhere.

File: python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py [hunk 7]
```diff
@@ -73,41 +105,157 @@ def __init__(self, **kwargs):
...
+        os.environ["NUMEXPR_MAX_THREADS"] = "1"
+
+    def getRefObjLoader(self, refCatalogList):
+        """
+        Create a `ReferenceObjectLoader` from available reference catalogs
+        in the repository.
+
+        Parameters
+        ----------
+        refCatalogList : list
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py]: refCatalogList : list -> refCatalogList : `list`

File: python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py [hunk 8]
```diff
@@ -73,41 +105,157 @@ def __init__(self, **kwargs):
...
+            List of deferred butler references for the reference catalogs.
+
+        Returns
+        -------
+        lsst.meas.algorithms.ReferenceObjectsLoader
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py]: lsst.meas.algorithms.ReferenceObjectsLoader -> `lsst.meas.algorithms.ReferenceObjectsLoader`

File: python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py [hunk 9]
```diff
@@ -73,41 +105,157 @@ def __init__(self, **kwargs):
...
+            config=LoadReferenceObjectsConfig(),
+        )
+        # This removes the padding around the border of detector BBox when
+        # matching to reference catalog.
+        # We remove this since we only want sources within detector.
+        refObjLoader.config.pixelMargin = 0
+
+        return refObjLoader
+
+    def runSelection(self, refObjLoader, bbox, wcs, filterName):
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py:144]: Do you want to do `filterName=None` based on your doc string in this function?

File: python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py [hunk 10]
```diff
@@ -73,41 +105,157 @@ def __init__(self, **kwargs):
...
+            return refSelection.sourceCat
+
+    def donutCatalogToDataFrame(self, donutCatalog, detectorName):
+        """
+        Reformat afwCatalog into a pandas dataframe sorted by flux with
+        the brightest objects at the top.
+
+        Parameters
+        ----------
+        donutCatalog : lsst.afw.table.SimpleCatalog
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py]: donutCatalog : lsst.afw.table.SimpleCatalog --> donutCatalog : `lsst.afw.table.SimpleCatalog`

File: python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py [hunk 11]
```diff
@@ -73,41 +105,157 @@ def __init__(self, **kwargs):
...
+            lsst.afw.table.SimpleCatalog object returned by the
+            ReferenceObjectLoader search over the detector footprint.
+        detectorName : str
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py]: detectorName : str -> detectorName : `str`

File: python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py [hunk 12]
```diff
@@ -73,41 +105,157 @@ def __init__(self, **kwargs):
...
+            Name of the detectors associated with donutCatalog.
+
+        Returns
+        -------
+        pandas DataFrame
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py]: `pandas.DataFrame` I guess?

File: python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py [hunk 13]
```diff
@@ -73,41 +105,157 @@ def __init__(self, **kwargs):
...
+        """
+
+        ra = []
+        dec = []
+        centroidX = []
+        centroidY = []
+        sourceFlux = []
+        detNames = []
+
+        if donutCatalog is not None:
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py:211]: Why do you have this judgement? The doc string in this function does not say the `donutCatalog` can be `None`. Right? Or you want to do the `donutCatalog=None` in the function input arguments and put the `optional` to the doc string?

**Review by teweitsai — APPROVED** [review]
The PR looks good to me. Please consider the comments before the merge. Thanks!

File: python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py (hunk 6 above)
Comment (jbkalmbach) [reply python/lsst/ts/wep/task/GenerateDonutCatalogOnlineTask.py]: Yes, I will use that instead.

File: python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py (hunk 9 above)
Comment (jbkalmbach) [reply python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py:144]: Actually, I will change the docstring. I think this must have been written before the code was in its final state and I forgot to update the docstring.

File: python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py (hunk 13 above)
Comment (jbkalmbach) [reply python/lsst/ts/wep/task/GenerateDonutCatalogWcsTask.py:211]: This should be optional. I will change the docstring and the order of the arguments.

**Pull request merged**